pymeasure/display/curves.py

```diff
@@ -98,14 +98,19 @@ def update_data(self):
         # use numpy array operations to arrange the data more efficiently
         data = self.results.data
         z_data = data[self.z].to_numpy(dtype=np.float64)
-
-        self.img_data.ravel()[: z_data.size] = z_data
-        print(self.img_data.shape, z_data.shape, self.xsize * self.ysize)
+        x_data = data[self.x].to_numpy(dtype=np.float64)
+        y_data = data[self.y].to_numpy(dtype=np.float64)
 
         # set image data
         if len(z_data) != 0:
+            x_unique = np.unique(x_data)
+            y_unique = np.unique(y_data)
+            x_ids = np.searchsorted(x_unique, x_data)
+            y_ids = np.searchsorted(y_unique, y_data)
```

Comment:
To prevent an awkward sorting of uncomplete data, you can use digitize rather than unique+searchsorted: (the bins are calculated in a separate suggestion).

```suggestion
            x_ids = np.digitize(x_data, self.xbins, right=True)
            y_ids = np.digitize(y_data, self.ybins, right=True)
```